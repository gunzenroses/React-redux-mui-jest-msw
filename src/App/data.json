{
  "todoList": [
    {
      "id": 1,
      "checked": false,
      "text": "Тестовое задание"
    },
    {
      "id": 2,
      "checked": true,
      "text": "Прекрасный код"
    },
    {
      "id": 3,
      "checked": false,
      "text": "Покрытие тестами"
    }
  ]
}