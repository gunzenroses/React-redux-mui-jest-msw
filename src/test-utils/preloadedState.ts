const initialState = [
  {
    id: 2,
    checked: false,
    text: 'first text',
  },
  {
    id: 5,
    checked: false,
    text: 'second text',
  },
];

export { initialState };