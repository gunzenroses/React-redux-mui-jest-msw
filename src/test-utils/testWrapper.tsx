import { PropsWithChildren, ReactElement } from "react";
import { Provider } from "react-redux";
import { configureStore, PreloadedState } from "@reduxjs/toolkit"
import { render, RenderOptions } from "@testing-library/react";

import todoSlice from "../redux/todoSlice";
import { MyState, MyStoreType } from '../redux/types';

interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: PreloadedState<MyState>,
  store?: MyStoreType
}

const setupStore = (preloadedState?: PreloadedState<MyState>) => {
  return configureStore({
    reducer: {
      todoList: todoSlice,
    },
    preloadedState,
  });
}

export const renderWithProviders = (
  ui: ReactElement,
  renderOptions?: ExtendedRenderOptions
) => {
  const store = setupStore({todoList: []});
  function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
    return <Provider store={store}>{children}</Provider>;
  }

  return render(ui, { wrapper: Wrapper, ...renderOptions });
};