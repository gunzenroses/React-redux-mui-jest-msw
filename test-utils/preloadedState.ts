const preloadedState = {
  todoList: [
    {
      id: 2,
      checked: false,
      text: 'test code',
    },
    {
      id: 5,
      checked: false,
      text: 'test code',
    },
  ],
};

export { preloadedState }