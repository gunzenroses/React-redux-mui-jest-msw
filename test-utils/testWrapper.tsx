import React, { PropsWithChildren, ReactElement } from "react";
import { Provider } from "react-redux";
import { configureStore, PreloadedState } from "@reduxjs/toolkit"
import { render, RenderOptions } from "@testing-library/react";

import todoSlice from "../src/redux/todoSlice";
import { MyState, MyStoreType } from '../src/redux/types';

interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: PreloadedState<MyState>,
  store?: MyStoreType
}

const setupStore = (preloadedState?: PreloadedState<MyState>) => {
  return configureStore({
    reducer: {
      todoList: todoSlice,
    },
    preloadedState,
  });
}

export const renderWithProviders = (
  ui: ReactElement, 
  {
    preloadedState = {todoList: []},
    store = setupStore(preloadedState),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) => {
  function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
    return <Provider store={store}>{children}</Provider>
  }

  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}