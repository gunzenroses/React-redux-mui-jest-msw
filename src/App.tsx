import { useState } from 'react';

import { Footer, Header, Panel, TodoList } from './components'
import './App.css';

import data from './components/data.json';

function App() {
  const [todoList, setTodoList] = useState(data.todoList);

  const [mode, setMode] = useState('All' as ModeType);

  const onDeleteItem = (id: TodoItemType['id']) => {
    setTodoList(todoList.filter((item) => item.id !== id));
  }

  const onAddItem = (text: string) => {
    setTodoList([
      ...todoList,
      {
        id: todoList[todoList.length - 1].id + 1,
        checked: false,
        text: text
      }
    ])
  }

  const onTextChange = (data: Omit<TodoItemType, 'checked'>) => {
    setTodoList(
      todoList.map((todo) => {
        if (todo.id === data.id) {
          return {
            ...data,
            checked: todo.checked,
          };
        }
        return todo;
      })
    );
  };

  const onCompleteChange = (data: Omit<TodoItemType, 'text'>) => {
    setTodoList(
      todoList.map((todo) => {
        if (todo.id === data.id) {
          return {
            ...data,
            text: todo.text,
          };
        }
        return todo;
      })
    );
  }

  const onModeChangeHandler = (mode: ModeType) => {
    setMode(mode);
  };

  const onDeleteCompletedHandler = () => {
    setTodoList(
      todoList.filter(todo => !todo.checked)
    )
  }

  return (
    <div className='App'>
      <Header />
      <Panel onAdd={onAddItem}/>
      <TodoList 
        todoList={todoList} 
        onDelete={onDeleteItem} 
        onTextChange={onTextChange}
        onCompletedChange={onCompleteChange}
      />
      <Footer
        listLength={todoList.length} 
        currentMode={mode}
        onModeChange={onModeChangeHandler}
        onDeleteCompleted={onDeleteCompletedHandler}
      />
    </div>
  );
}

export default App;











