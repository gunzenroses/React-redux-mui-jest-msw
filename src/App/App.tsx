import { useEffect, useMemo, useState } from 'react';

import { useMyDispatch, useMyList } from '../redux/hooks';
import { getTodos, changeTodo, deleteTodo } from '../redux/todoThunk';
import { Footer, Header, Panel, TodoList } from '../components';
import './App.css';

function App() {
  const dispatch = useMyDispatch();

  const data: TodoItemType[] = useMyList();
  console.log(1, data)

  useEffect(() => {
    dispatch(getTodos());
  }, []);

  const onDeleteCompletedHandler = () => {
    const completedList = data.filter((todo) => todo.checked);
    if (completedList.length < 1) return;
    const completedIds: TodoItemType['id'][] = completedList.map((todo) => todo.id);
    dispatch(deleteTodo(completedIds));
  }

  const [mode, setMode] = useState('All' as ModeType);

  const onModeChangeHandler = (mode: ModeType) => {
    setMode(mode);
  };

  const [modList, setModList] = useState(data);

  useEffect(() => {
    const modifiedList = data.filter(todo => {
      switch(mode) {
        case 'All': return todo;
        case 'Active': return !todo.checked;
        case 'Completed': return todo.checked;
      }
    });
    setModList(modifiedList);
  }, [mode, data]);

  return (
    <div className='App'>
      <Header />
      <Panel lastDataId={data.length > 0 ? (data[data.length-1].id + 1) : 1} />
      <TodoList
        todoList={modList}
        currentMode={mode}
      />
      <Footer
        listLength={modList.length}
        currentMode={mode}
        onModeChange={onModeChangeHandler}
        onDeleteCompleted={onDeleteCompletedHandler}
      />
    </div>
  );
}

export default App;
































