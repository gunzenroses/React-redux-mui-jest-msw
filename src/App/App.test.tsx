import { fireEvent, screen, waitFor } from '@testing-library/react';

import { initialState } from '../test-utils/preloadedState';
import { renderWithProviders } from '../test-utils/testWrapper';
import App from './App';

test('App should match snapshot', async () => {
  const { container } = renderWithProviders(<App />);

  await waitFor(() => screen.getAllByLabelText('todoItem'));

  expect(container).toMatchSnapshot();
});

describe('button Add new todo', () => {
  test('works when smth typed it', async () => {
    const newText = 'something new';

    renderWithProviders(<App />);

    await waitFor(() => screen.getAllByLabelText('todoItem'));

    const inputTodo = screen.getByPlaceholderText("What's need to be done?") as HTMLInputElement;
    const addButton = screen.getByLabelText(/addButton/i);
    
    await waitFor(() => {
      fireEvent.change(inputTodo, { target: { value: newText } });
    });

    expect(inputTodo.value).toBe(newText);

    await waitFor(() => {
      fireEvent.click(addButton);
    })

    expect(inputTodo.value).toBe('');
  })
})

describe('Change todoItem',  () => {
  const newText = 'this is it';

  test('should work after pushing the change button', async () => {
    renderWithProviders(<App />);
    await waitFor(() => screen.getAllByLabelText('todoItem'));
    const firstTodo = screen.getByDisplayValue(initialState[0].text) as HTMLInputElement;

    const changeBtn = screen.getAllByLabelText('changeBtn')[0];
    await waitFor(() => {
      fireEvent.click(changeBtn);
    }).then(() => {
      fireEvent.change(firstTodo, { target: { value: newText } });
    });
 
    expect(firstTodo.value).toBe(newText);
  })

  test('should not work when change button is not pushed', async () => {
    renderWithProviders(<App />);

    await waitFor(() => screen.getAllByLabelText('todoItem'));

    const firstTodo = screen.getByDisplayValue(
      initialState[0].text
    ) as HTMLInputElement;
    await waitFor(() => {
      fireEvent.change(firstTodo, { target: { value: newText } });
    });

    expect(firstTodo).not.toHaveDisplayValue(newText);
  })
})

describe('Delete todoItem', () => {
  test('should work', async () => {
    const firstValue = initialState[0].text;
    renderWithProviders(<App />),
    await waitFor(() => { screen.getAllByLabelText('todoItem')});
    const firstTodo = screen.getByDisplayValue(firstValue);
    const deleteBtn = screen.getAllByLabelText('deleteBtn')[0];

    expect(firstTodo).toBeInTheDocument();

    await waitFor(() => {
      fireEvent.click(deleteBtn);
    });
    console.log('should be after 4');

    const newFirstTodo = screen
      .getAllByLabelText('todoItem')[0]
      .getElementsByTagName('input')[0].value;
    console.log(newFirstTodo);

    expect(firstTodo).not.toBeInTheDocument();
  });
})

describe('', () => {

})