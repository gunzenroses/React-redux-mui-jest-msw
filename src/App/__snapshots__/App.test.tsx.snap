// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match snapshot 1`] = `
<div>
  <div
    class="App"
  >
    <h1
      class="MuiTypography-root MuiTypography-body1 MuiTypography-alignCenter css-1i7zwci-MuiTypography-root"
    >
      Todo List
    </h1>
    <div
      class="MuiBox-root css-cllfw7"
    >
      <div
        class="MuiFormControl-root MuiTextField-root css-w3s1g-MuiFormControl-root-MuiTextField-root"
      >
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
            id=":r0:"
            placeholder="What's need to be done?"
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-ihdtdm"
            >
              <span
                class="notranslate"
              >
                â€‹
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <button
        aria-label="addButton"
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium Mui-disabled MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-axubsc-MuiButtonBase-root-MuiButton-root"
        disabled=""
        tabindex="-1"
        type="button"
      >
        <span
          class="MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="AddIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
            />
          </svg>
        </span>
        Add
      </button>
    </div>
    <div
      class="MuiBox-root css-1v8my8o"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation2 css-czvv85-MuiPaper-root"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-colorError MuiSvgIcon-fontSizeLarge css-1o657us-MuiSvgIcon-root"
          data-testid="RadioButtonUncheckedIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
          />
        </svg>
        <div
          aria-label="todoItem"
          class="MuiFormControl-root MuiTextField-root css-zwh86o-MuiFormControl-root-MuiTextField-root"
        >
          <div
            class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-formControl css-1ptx2yq-MuiInputBase-root-MuiInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
              id=":r1:"
              type="text"
              value="first text"
            />
          </div>
        </div>
        <button
          aria-label="changeBtn"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-lwmgk2-MuiButtonBase-root-MuiIconButton-root"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge css-tzssek-MuiSvgIcon-root"
            data-testid="ModeEditIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 0 0-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
            />
          </svg>
        </button>
        <button
          aria-label="deleteBtn"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-13b8k9w-MuiButtonBase-root-MuiIconButton-root"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge css-tzssek-MuiSvgIcon-root"
            data-testid="DeleteIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
            />
          </svg>
        </button>
      </div>
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation2 css-czvv85-MuiPaper-root"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-colorError MuiSvgIcon-fontSizeLarge css-1o657us-MuiSvgIcon-root"
          data-testid="RadioButtonUncheckedIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
          />
        </svg>
        <div
          aria-label="todoItem"
          class="MuiFormControl-root MuiTextField-root css-zwh86o-MuiFormControl-root-MuiTextField-root"
        >
          <div
            class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-formControl css-1ptx2yq-MuiInputBase-root-MuiInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
              id=":r2:"
              type="text"
              value="second text"
            />
          </div>
        </div>
        <button
          aria-label="changeBtn"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-lwmgk2-MuiButtonBase-root-MuiIconButton-root"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge css-tzssek-MuiSvgIcon-root"
            data-testid="ModeEditIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 0 0-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
            />
          </svg>
        </button>
        <button
          aria-label="deleteBtn"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-13b8k9w-MuiButtonBase-root-MuiIconButton-root"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge css-tzssek-MuiSvgIcon-root"
            data-testid="DeleteIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
            />
          </svg>
        </button>
      </div>
    </div>
    <div
      class="MuiBox-root css-1jqegk3"
    >
      <p
        class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
      >
        2
         
        items
         left
      </p>
      <div
        class="MuiBox-root css-kl5uk3"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-10n65wt-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          All
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium css-dn2g4c-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          Active
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium css-dn2g4c-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          Completed
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedError MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedError MuiButton-sizeMedium MuiButton-containedSizeMedium css-1sszw6i-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        Clear completed
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</div>
`;
